File: rust/portainer-lib/src/output.rs
use crate::error::PortainerError;
use serde::Serialize;

#[derive(Serialize)]
pub struct ErrorOutput {
    pub error: ErrorDetail,
}

#[derive(Serialize)]
pub struct ErrorDetail {
    pub code: String,
    pub message: String,
}

pub fn print_yaml<T: Serialize>(data: &T) -> Result<(), PortainerError> {
    let yaml = serde_yaml::to_string(data)
        .map_err(|e| PortainerError::ApiError(format!("Failed to serialize: {}", e)))?;
    print!("{}", yaml);
    Ok(())
}

pub fn print_error(err: &PortainerError) {
    let output = ErrorOutput {
        error: ErrorDetail {
            code: err.code().to_string(),
            message: err.to_string(),
        },
    };
    if let Ok(yaml) = serde_yaml::to_string(&output) {
        eprint!("{}", yaml);
    } else {
        eprintln!("error: {}", err);
    }
}

File: trans/pkg/trans/output.go
package trans

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

type ErrorOutput struct {
	Error ErrorDetail `yaml:"error"`
}

type ErrorDetail struct {
	Code    string `yaml:"code"`
	Message string `yaml:"message"`
}

func PrintYAML(data interface{}) error {
	enc := yaml.NewEncoder(os.Stdout)
	enc.SetIndent(2)
	if err := enc.Encode(data); err != nil {
		return APIError(fmt.Sprintf("failed to serialize: %s", err))
	}
	return nil
}

func PrintError(err error) {
	te, ok := err.(*TransError)
	if !ok {
		fmt.Fprintf(os.Stderr, "error: %s\n", err)
		return
	}

	output := ErrorOutput{
		Error: ErrorDetail{
			Code:    string(te.Code),
			Message: te.Message,
		},
	}

	enc := yaml.NewEncoder(os.Stderr)
	enc.SetIndent(2)
	if err := enc.Encode(output); err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", te.Message)
	}
}
